File: src/main/java/com/EduPay/model/Payment.java
package com.EduPay.model;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;

import java.time.LocalDateTime; // For payment date and time

@Entity
@Table(name = "payments") // Specify table name for clarity (plural is common)
@Data // Lombok annotation for getters, setters, toString, equals, hashCode
@NoArgsConstructor // Lombok annotation for no-argument constructor
@AllArgsConstructor // Lombok annotation for all-argument constructor
public class Payment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id; // Primary key for the payment record

    @Column(nullable = false)
    private String transactionId; // Unique ID for the transaction (e.g., from Razorpay or internal)

    @Column(nullable = false)
    private Double amount; // Amount of the payment

    @Column(nullable = false)
    private String paymentMethod; // "Cash" or "Online"

    @Column(nullable = false)
    private LocalDateTime paymentDate; // Date and time when the payment was made

    @Column(nullable = false)
    private String status; // e.g., "Success", "Pending", "Failed"

    // For online payments, store gateway-specific details
    @Column(nullable = true)
    private String gatewayPaymentId; // e.g., Razorpay payment ID

    @Column(nullable = true)
    private String gatewayOrderId; // e.g., Razorpay order ID

    // Relationship with Student entity: Many payments belong to one student
    @ManyToOne
    @JoinColumn(name = "student_id", nullable = false) // Foreign key column
    private Student student; // The student who made this payment

    // If cash payment, this links to the admin (User entity) who recorded it
    @ManyToOne
    @JoinColumn(name = "recorded_by_user_id", nullable = true) // Nullable because online payments aren't 'recorded' by an admin in this way
    private User recordedBy; // The admin user who manually recorded this cash payment


}

File: frontend/edupay/lib/models/payment.dart
// lib/models/payment.dart

import 'dart:convert';

class Payment {
  final int? id;
  final String transactionId;
  final double amount;
  final String paymentMethod;
  final DateTime paymentDate;
  final String status;
  final String? gatewayPaymentId;
  final String? gatewayOrderId;
  final int? studentId; // Link to the student
  final int? recordedById; // Link to the admin who recorded it (for cash)

  Payment({
    this.id,
    required this.transactionId,
    required this.amount,
    required this.paymentMethod,
    required this.paymentDate,
    required this.status,
    this.gatewayPaymentId,
    this.gatewayOrderId,
    this.studentId,
    this.recordedById,
  });

  // Create a Payment object from a JSON map
  factory Payment.fromJson(Map<String, dynamic> json) {
    return Payment(
      id: json['id'] as int?,
      transactionId: json['transactionId'] as String,
      amount: json['amount'] as double,
      paymentMethod: json['paymentMethod'] as String,
      paymentDate: json['paymentDate'] is List
          ? DateTime(
              json['paymentDate'][0],
              json['paymentDate'][1],
              json['paymentDate'][2],
              json['paymentDate'].length > 3 ? json['paymentDate'][3] : 0,
              json['paymentDate'].length > 4 ? json['paymentDate'][4] : 0,
              json['paymentDate'].length > 5 ? json['paymentDate'][5] : 0)
          : DateTime.parse(json['paymentDate'].toString()),
      status: json['status'] as String,
      gatewayPaymentId: json['gatewayPaymentId'] as String?,
      gatewayOrderId: json['gatewayOrderId'] as String?,
      studentId: json['studentId'] as int?,
      recordedById: json['recordedById'] as int?,
    );
  }

  // Convert a Payment object to a JSON map
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'transactionId': transactionId,
      'amount': amount,
      'paymentMethod': paymentMethod,
      'paymentDate': paymentDate.toIso8601String(), // Send as ISO 8601 string
      'status': status,
      'gatewayPaymentId': gatewayPaymentId,
      'gatewayOrderId': gatewayOrderId,
      'studentId': studentId,
      'recordedById': recordedById,
    };
  }
}
// Note: The Payment class is designed